### PR TITLE
Support custom html translation extractors

### LINKED ISSUE #19: Using translate element reports missing translation ID
When using the translate directive in it's own tag, the tool reports a missing translation ID

```
<translate translate="TRANSLATION-ID"></translate>
```

Documentation is [here](https://angular-translate.github.io/docs/#/api/pascalprecht.translate.directive:translate)

### PR DESCRIPTION
**Pull request by MichaReiser**
Adds support for custom HTML text extractors and rewrites the AngularJS extractor to be based on the same API.

The API of an extractor is:

```
export interface HtmlTranslationExtractor {
  (element: AngularElement, context: HtmlTranslationExtractionContext): void;
}

export interface AngularElement {
  tagName: string;
  attributes: Attribute[];
  texts: Text[];
  startPosition: number;
}

export interface HtmlTranslationExtractionContext {
  emitError(message: string, position: number): void;
  emitSuppressableError(message: string, position: number): void;
  registerTranslation(translation: {
    translationId: string;
    defaultText?: string;
    position: number;
  }): void;

  asHtml(): void;
}
```

The extractor receives an angular element with all its attributes and its direct text siblings as well as a context that can be used to either register a new translation or emit a warning/error. 

The `translate-directive-translation-extractor.ts` contains an implementation of an extractor.

Custom extractors can be specified with the html loader:

```js
{
  loader: WebPackAngularTranslate.htmlLoader({
    translationExtractors: [(element, context) => { ... }]
  })
}
```

Steps required for finalization / open questions:

- Verify if this PR addresses the needs of #37 and #36 
- Verify if the assumptions that an angular HTML document always has at least one root element (it's not possible that a document only contains text nodes)
- Work out if it should be possible to remove the default translation extractor or if this one should be always added (is the extractors array additive or does it override the extractors)
- Code cleanup / Fix TODOs
- Mention in README

Fixes #19 #37 and  #36

### CODE REVIEW

**Comment by MartijnWelker** [comment]
Very clear implementation, looks like it will fit all the needs (and more !) for #36

File: src/html/translate-directive-translation-extractor.ts [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+function handleAngularExpression(
+  expression: AngularExpressionMatch,
+  context: HtmlTranslationExtractionContext,
+  position: number
+) {
+  let translationId = expression.value;
+  // e.g { translate | var} instead of a string constant
+  if (!(/^".*"$/.test(translationId) || /^.*'$/.test(translationId))) {
```
Comment (sambernet) [inline src/html/translate-directive-translation-extractor.ts]: Missing single quote in second regex I reckon:
```suggestion
  if (!(/^".*"$/.test(translationId) || /^'.*'$/.test(translationId))) {
```

File: src/html/translate-html-parser.ts [hunk 2]
```diff
@@ -21,212 +29,127 @@ function isAngularExpression(value: string): boolean {
...
+      this.context.emitSuppressableError(
         `The element '${this.context.asHtml()}'  in '${
           this.loader.resource
-        }' uses an angular expression as translation id ('${translationId}') or as default text ('${defaultText}'), this is not supported. To suppress this error at the '${SUPPRESS_ATTRIBUTE_NAME}' attribute to the element or any of its parents.`,
-        loc
+        }' uses an angular expression as translation id ('${
+          translation.translationId
+        }') or as default text ('${
+          translation.defaultText
+        }'), this is not supported. To suppress this error at the '${SUPPRESS_ATTRIBUTE_NAME}' attribute to the element or any of its parents.`,
```
Comment (sambernet) [inline src/html/translate-html-parser.ts]:
```suggestion
        }'). This is not supported. To suppress this error add the '${SUPPRESS_ATTRIBUTE_NAME}' attribute to the element or any of its parents.`,
```

File: src/html/html-translation-extractor.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  tagName: string;
+  attributes: Attribute[];
+  texts: Text[];
+  startPosition: number;
+}
+
+export interface HtmlTranslationExtractionContext {
+  emitError(message: string, position: number): void;
+  emitSuppressableError(message: string, position: number): void;
+  registerTranslation(translation: {
```
Comment (sambernet) [inline src/html/html-translation-extractor.ts]: The type of the `translation` parameter is used (and declared inline) several times, introducing a type alias would remove some duplication here (even if only in typings and not affecting generated JS):

```
export type Translation = {
    translationId: string;
    defaultText?: string;
    position: number;
}
```

Edit: I see this conflicts with the existing `Translation` class, but still defining a type (with a slightly different name, e.g. `TranslationUnit`) could be beneficial (and `Translation` class could implement it, too, if `id` is renamed to `translationId` in there also).

File: src/html/element-context.ts [hunk 4]
```diff
@@ -1,78 +1,186 @@
...
+   * The text contents of the element
+   */
+  readonly texts: Text[] = [];
+
+  public suppressDynamicTranslationErrors: boolean;
+
+  /**
+   * Is the translate directive applied to the current element (translate element or element with translate attribute)
+   */
+  translateDirective = false;
```
Comment (sambernet) [inline src/html/element-context.ts]: Property now obsolete?!

File: src/html/element-context.ts [hunk 5]
```diff
@@ -1,78 +1,186 @@
...
+
+  /**
+   * Does the element has translate-attr-* attributes?
+   */
+  translateAttributes = true;
```
Comment (sambernet) [inline src/html/element-context.ts]: Property now obsolete?!

File: src/html/element-context.ts [hunk 6]
```diff
@@ -1,78 +1,186 @@
...
+
+  /**
+   * The translation id
+   */
+  translationId: string;
```
Comment (sambernet) [inline src/html/element-context.ts]: Property now obsolete?!

File: src/html/element-context.ts [hunk 7]
```diff
@@ -1,78 +1,186 @@
...
+
+  /**
+   * The default text of the translation
+   */
+  defaultText: string;
```
Comment (sambernet) [inline src/html/element-context.ts]: Property now obsolete?!

**Comment by sambernet** [comment]
We tested the PR on our current angularJS project and did not find any difference in the translations extracted (test file has 700+ translations). ✅ Plugin performance was not yet compared.

Now working on an extractor for the Angular `i18n` elements according to #37. We do think this will satisfy all needs and will give an update here once ready.

Thanks a lot already, great job!

[OT]Using the brand-new github draft PRs. Always cutting-edge @MichaReiser  😄 💪 [/OT]

File: src/html/translate-directive-translation-extractor.ts (hunk 1 above)
Comment (MichaReiser) [reply src/html/translate-directive-translation-extractor.ts]: Nice catch!

**Comment by MichaReiser** [comment]
@sambernet Thanks for the detailed review! I addressed the comments and the one open todo so that I believe it should be ready to merge.

**Pull request merged**